lib/ansible/utils/singleton.py

```diff
@@ -0,0 +1,21 @@
+# Copyright (c) 2017 Ansible Project
+# GNU General Public License v3.0+ (see COPYING or https://www.gnu.org/licenses/gpl-3.0.txt)
+
+# Make coding more python3-ish
+from __future__ import (absolute_import, division, print_function)
+__metaclass__ = type
+
+
+class Singleton(type):
+    """Metaclass for classes that wish to implement Singleton
+    functionality.  If an instance of the class exists, it's returned,
+    otherwise a single instance is instantiated and returned.
+    """
+    def __init__(cls, name, bases, dct):
+        super(Singleton, cls).__init__(name, bases, dct)
+        cls.__instance = None
+
+    def __call__(cls, *args, **kw):
+        if cls.__instance is None:
+            cls.__instance = super(Singleton, cls).__call__(*args, **kw)
```

Comment:
*definitely* need a double-check lock on the inner case here for when we start hitting these with threads. Risk is likely very low today in the `Display` case, but this is a general piece of utility code I can see being used elsewhere.

Reply:
I've added a reentrant lock in this method, which should prevent deadlocks if mixing multiple Singletons in the future, but would be thread safe.